src/psyclone/psyir/frontend/fparser2.py

```diff
@@ -451,6 +451,62 @@ def _is_range_full_extent(my_range):
     return is_lower and is_upper and is_step
 
 
+def _create_array_bound_arg(node):
+    '''
+    Given the supplied node, creates a new node with the same access
+    apart from the final array access, suitable for use as an
+    argument to either LBOUND or UBOUND.
+
+    e.g. if `node` is an ArrayMember representing the inner access in
+    'grid%data(:)' then this routine will return a PSyIR node for
+    'grid%data'.
+
+    :param node: the array access. In the case of a structure, this \
+                 must be the inner-most part of the access.
+    :type node: :py:class:`psyclone.psyir.nodes.Reference` or \
+                :py:class:`psyclone.psyir.nodes.Member`
+
+    :returns: the PSyIR for a suitable argument to either LBOUND or \
+              UBOUND applied to the supplied `node`.
+    :rtype: :py:class:`psyclone.psyir.nodes.Node`
+
+    :raises InternalError: if the supplied node is not an instance of \
+                           either Reference or Member.
+    '''
+    if isinstance(node, Reference):
+        return Reference(node.symbol)
+
+    elif isinstance(node, Member):
+        # We have to take care with derived types:
+        # grid(1)%data(:...) becomes
+        # grid(1)%data(lbound(grid(1)%data,1):...)
+        # N.B. the argument to lbound becomes a Member access rather
+        # than an ArrayMember access.
+        parent_ref = node.ancestor(Reference, include_self=True)
+        # We have to find the location of the supplied node in the
+        # StructureReference.
+        inner = parent_ref
+        depth = 0
+        while hasattr(inner, "member") and inner is not node:
+            depth += 1
+            inner = inner.member
+        # Now we take a copy of the full reference and then modify it so
+        # that the copy of 'node' is replaced by a Member().
+        arg = parent_ref.copy()
+        # We use the depth computed for the original reference in order
+        # to find the copy of 'node'.
+        inner = arg
+        for step in range(depth-1):
+            inner = inner.member
+        # Change the innermost access to be a Member.
+        inner.children[0] = Member(node.name, inner)
+        return arg
+
+    raise InternalError(
+        "The supplied node must be an instance of either ArrayReference "
```

Comment:
Looking at the conditions I guess this should be "either a Reference or a Member"

Reply:
Oops again.